src/sentry/notifications/notifications/mixins/role_based_mixin.py

```diff
@@ -0,0 +1,58 @@
+from __future__ import annotations
+
+from typing import TYPE_CHECKING, Iterable, MutableMapping
+
+from sentry import roles
+from sentry.models import OrganizationMember, Team
+
+if TYPE_CHECKING:
+    from sentry.models import Organization, User
+
+
+class RoleBasedMixin:
+    member_by_user_id: MutableMapping[int, OrganizationMember] = {}
+    organization: Organization
+
+    def get_member(self, user: User) -> OrganizationMember:
+        # cache the result
+        if user.id not in self.member_by_user_id:
+            self.member_by_user_id[user.id] = OrganizationMember.objects.get(
+                user=user, organization=self.organization
+            )
+        return self.member_by_user_id[user.id]
+
+    def set_member_in_cache(self, member: OrganizationMember) -> None:
+        """
+        A way to set a member in a cache to avoid a query.
+        """
+        self.member_by_user_id[member.user_id] = member
+
+    def determine_recipients(self) -> Iterable[Team | User]:
```

Comment:
I'm not sure that having this method which also exists on `BaseNotification` is an anti-pattern